include/seastar/util/noncopyable_function.hh

```diff
@@ -66,10 +67,11 @@ private:
         // that, and complains. Silence it.
 #pragma GCC diagnostic push
 #pragma GCC diagnostic ignored "-Wuninitialized"
-        // Avoid including <algorithm> just for this
-        for (unsigned i = 0; i != N; ++i) {
-            to->_storage.direct[i] = from->_storage.direct[i];
-        }
+        // A bytewise loop doesn't optimize to a wordwise loop since
+        // the compiler can't guess if there is overlap or not. Help
+        // it out by using a memcpy. This translates to wordwise
+        // copies.
+        std::memcpy(to->_storage.direct, from->_storage.direct, N);
```

Comment:
Looks good to me, but I thought over the years that you tried to avoid std::memcpy and use std::copy_n instead? Or you like both?

Reply:
I prefer std::copy_n(), but here the optimizer is too careful, as I explained in the commit message. std::copy_n() cannot be optimized if the optimizer thinks overlap is possible.

Reply:
According to the documentation of copy_n, "Overlap of ranges is formally permitted, but leads to unpredictable ordering of the results." - so the compiler can do anything if it thinks overlaps are possible (and in particular, use the optimized code). But maybe it indeed doesn't optimize - I didn't check.

In any case, I have nothing against std::memcpy - just that I'm still surprised about the "hutzpah" of compiler writers to convert explicit calls to library functions by their own builtin functions.

Reply:
I think memcpy is less fragile here, copy_n depends too much on the library internals (which eventually forward the call to memcpy).